### PR TITLE
Add defer to docs and compile commands

### LINKED ISSUE #4110: [Feature] dbt snapshot --defer
### Is there an existing feature request for this?

- [X] I have searched the existing issues

### Describe the Feature

Implement "dbt snapshot --defer", which would behave similar to "dbt run --defer".

### Describe alternatives you've considered

_No response_

### Who will this benefit?

When trying to set up "slim ci", one step in my case would be to run snapshots based on other dbt models.

`dbt snapshot --select 'dbt.*,state:modified' --state manifests/staging`

When running this command, generated references point to current schema, which doesn't have these tables, as they weren't generated during preceding run command:
`dbt run --models state:modified --defer --state manifests/airflow/staging`

### Are you interested in contributing this feature?

_No response_

### Anything else?

https://docs.getdbt.com/reference/node-selection/defer
https://docs.getdbt.com/docs/guides/best-practices#run-only-modified-models-to-test-changes-slim-ci

dbt version:  0.20.1

### PR DESCRIPTION
**Pull request by kadero**
resolves #4110 #3741

### Checklist

- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [ ] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have updated the `CHANGELOG.md` and added information about my change

### CODE REVIEW

**Comment by ChenyuLInx** [comment]
Thanks for contributing this @kadero! Sorry for getting back to you late. Can we add some tests to validate the added functionality?

**Comment by kadero** [comment]
> Thanks for contributing this @kadero! Sorry for getting back to you late. Can we add some tests to validate the added functionality?

Hi @ChenyuLInx, 
Thanks for your message :slightly_smiling_face:,
I will work on those tests on it in the next few days :+1:

**Comment by kadero** [comment]
> > Thanks for contributing this @kadero! Sorry for getting back to you late. Can we add some tests to validate the added functionality?
> 
> Hi @ChenyuLInx, Thanks for your message slightly_smiling_face, I will work on those tests on it in the next few days +1

@ChenyuLInx done :heavy_check_mark:

**Comment by ChenyuLInx** [comment]
> @ChenyuLInx done ✔️
Thanks for adding the tests!! One question regarding this test and the one you added for the snapshot. Is there any way that we can validate one step further that the defer actually used the previous state?(Looks like it did since without providing the previous state, the run would fail) Maybe we can validate more things returned in the defer run?

Also I should have noticed it the first time but as I was going over Jeremy's comment in #4110, seems like we can 
* Move `defer_to_manifest` up from RunTask into [CompileTask](https://github.com/dbt-labs/dbt-core/blob/f79a968a0908add9de1b6acfdb75a6b205134ebd/core/dbt/task/compile.py#L36) (its parent)

With this change you put in right? that would make things cleaner?

**Comment by kadero** [comment]
> > @ChenyuLInx done heavy_check_mark
> > Thanks for adding the tests!! One question regarding this test and the one you added for the snapshot. Is there any way that we can validate one step further that the defer actually used the previous state?(Looks like it did since without providing the previous state, the run would fail) Maybe we can validate more things returned in the defer run?
> 
> Also I should have noticed it the first time but as I was going over Jeremy's comment in #4110, seems like we can
> 
> * Move `defer_to_manifest` up from RunTask into [CompileTask](https://github.com/dbt-labs/dbt-core/blob/f79a968a0908add9de1b6acfdb75a6b205134ebd/core/dbt/task/compile.py#L36) (its parent)
> 
> With this change you put in right? that would make things cleaner?

Hi @ChenyuLInx :wave:,

Thank you for your comment. It's done :heavy_check_mark: `defer_to_manifest` is moved to CompileTask.

Please tell me if something is missing in this PR :slightly_smiling_face:

:pray:

File: test/integration/029_docs_generate_tests/test_docs_generate.py [hunk 1]
```diff
@@ -2107,3 +2107,32 @@ def test_postgres_long_paths(self):
...
+
+    @use_profile('postgres')
+    def test_postgres_generate_defer(self):
+        self.assertEqual(len(self.run_dbt(['run'])), 1)
+
+        # copy state files
+        self.copy_state()
+
+        # defer test, it succeeds
+        results = self.run_dbt(['docs', 'generate', '--state', 'state', '--defer'])
```
Comment (ChenyuLInx) [inline test/integration/029_docs_generate_tests/test_docs_generate.py:2138]: Same here, can we check different behavior of defer vs no defer?

**Review by ChenyuLInx — CHANGES_REQUESTED** [review]
Hey @kadero , sorry for getting back late! Thanks for moving the previous part code around. I somehow missed to add the two comments I added this time. Can we be a bit more specific on the tests so that it doesn't only check that we can provide defer command but also check that defer works as expected under the hood.

**Comment by katieclaiborne** [comment]
Is there a similar issue related to the `dbt ls` command?

**Comment by ChenyuLInx** [comment]
> Is there a similar issue related to the `dbt ls` command?

@katieclaiborne  Sorry for getting back to you late. Are you asking that is there an issue related to also adding defer for `ls` command?

File: test/integration/062_defer_state_tests/test_defer_state.py [hunk 2]
```diff
@@ -47,6 +47,22 @@ def copy_state(self):
...
+        assert len(results) == 2
+        assert not any(r.node.deferred for r in results)
+        results = self.run_dbt(['test'])
+        assert len(results) == 2
+
+        # copy files
+        self.copy_state()
+
+        # defer test, it succeeds
+        self.run_dbt(['compile', '--state', 'state', '--defer'])
```
[the remark this answers is not in the document]
Comment (kadero) [reply test/integration/062_defer_state_tests/test_defer_state.py]: Added :heavy_check_mark:. Please tell me @ChenyuLInx if you want to assert something more specific.

:pray:

**Comment by katieclaiborne** [comment]
> Are you asking that is there an issue related to also adding defer for ls command?

Yes, exactly!

**Review by ChenyuLInx — APPROVED** [review]
Looks good to me! @kadero  thanks for adding this and the test!

**Comment by ChenyuLInx** [comment]
> > Are you asking that is there an issue related to also adding defer for ls command?
> 
> Yes, exactly!

@katieclaiborne  I don't think so, feel free to open an issue about it!

**Comment by katieclaiborne** [comment]
Oops, I just realized that this is already possible with [the state method](https://docs.getdbt.com/reference/node-selection/methods#the-state-method)!

**Pull request merged**